File: dev_tools/git_history.js
#!/usr/bin/env node
/**
 * Git SLOC History Visualizer
 *
 * This CLI tool goes through your git history (in chronological order),
 * and for each commit it does the following:
 *
 *  1. Exports the commit’s files to a temporary directory.
 *  2. Runs three series of SLOC counts using npx node-sloc:
 *      - Series 1: All code in the “lib/” folder.
 *      - Series 2: Non-testing library code (ignoring **\*.test.ts).
 * - Series 3: Effective compiled code in “out /” (ignoring **\*.test.js).
 *  3. Parses the SLOC output from each command.
 *  4. Aggregates the commit’s date, hash, and SLOC counts.
 *
 * Finally, it produces a polished HTML file (with Plotly.js) that you can open in your browser.
 *
 * This tool is written as an ES6 module for NodeJS and is Windows‐friendly.
 */

import { exec as execCb } from "child_process";
import { promisify } from "util";
import fs from "fs";
import path from "path";
import os from "os";
import * as tar from "tar";

const exec = promisify(execCb);

/**
 * Retrieve the full list of commits (chronological order).
 * Uses "git log" with a custom format: <commit hash> <commit timestamp>.
 */
async function getCommits() {
  const { stdout } = await exec('git log --reverse --pretty=format:"%H %ct"');
  return stdout
    .trim()
    .split("\n")
    .map((line) => {
      const [hash, timestamp] = line.split(" ");
      return { hash, timestamp: Number(timestamp) };
    });
}

/**
 * Run the node-sloc command with the given arguments in the provided working directory.
 * Returns the SLOC number (or 0 on error or if not found).
 */
async function runNodeSloc(cwd, argsArray) {
  // Build command string, e.g.,
  // npx node-sloc lib/
  // npx node-sloc lib/ --ignore-paths "**/*.test.ts"
  const command = `npx node-sloc ${argsArray.join(" ")}`;
  try {
    const { stdout } = await exec(command, { cwd });
    // Expect a table row like: | SLOC                          | 925               |
    const match = stdout.match(/\|\s*SLOC\s*\|\s*(\d+)/);
    if (match) {
      return Number(match[1]);
    }
  } catch (err) {
  }
  return 0;
}

/**
 * Process a single commit:
 *   - Create a temporary directory.
 *   - Archive and extract the commit into it.
 *   - Run the three node-sloc commands.
 *   - Clean up and return an object with commit info and SLOC numbers.
 */
async function processCommit(commit) {
  // Create a unique temporary directory
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "git-sloc-"));
  const archivePath = path.join(tmpDir, "archive.tar");

  try {
    // Archive the commit to a tar file.
    await exec(`git archive ${commit.hash} --format=tar -o "${archivePath}"`);
  } catch (err) {
    console.error(`Error archiving commit ${commit.hash}:`, err);
    return null;
  }

  try {
    // Extract the tar archive into the temporary directory.
    await tar.extract({ file: archivePath, cwd: tmpDir });
  } catch (err) {
    console.error(`Error extracting commit ${commit.hash}:`, err);
    return null;
  }

  // Run the three series of node-sloc commands.
  const series1 = await runNodeSloc('.', [tmpDir + "/lib/"]);
  const series2 = await runNodeSloc('.', [tmpDir + "/lib/", "--ignore-paths", `"**/*.test.ts"`]);
  const series3 = await runNodeSloc('.', [tmpDir + "/out/", "--ignore-paths", `"**/*.test.js"`]);

  // Clean up the temporary directory.
  fs.rmSync(tmpDir, { recursive: true, force: true });

  return { ...commit, series1, series2, series3 };
}

/**
 * Generate an HTML file with an embedded Plotly.js chart.
 * The chart shows three lines (one for each series) versus commit date.
 */
function generateHTML(data) {
  // Prepare arrays for dates, SLOC counts, and commit hashes (for hover text).
  const dates = data.map((d) => new Date(d.timestamp * 1000).toISOString().split("T")[0]);
  const series1 = data.map((d) => d.series1);
  const series2 = data.map((d) => d.series2);
  const series3 = data.map((d) => d.series3);
  const commitHashes = data.map((d) => d.hash);

  return `
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Git History SLOC Progression</title>
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
    <style>
      body { font-family: Arial, sans-serif; margin: 40px; }
      h1 { text-align: center; }
      #chart { width: 100%; height: 600px; }
    </style>
  </head>
  <body>
    <h1>Git History SLOC Progression</h1>
    <div id="chart"></div>
    <script>
      var dates = ${JSON.stringify(dates)};
      var series1 = ${JSON.stringify(series1)};
      var series2 = ${JSON.stringify(series2)};
      var series3 = ${JSON.stringify(series3)};
      var commitHashes = ${JSON.stringify(commitHashes)};

      var trace1 = {
        x: dates,
        y: series1,
        mode: 'lines+markers',
        name: 'All Code (lib/)',
        text: commitHashes,
        hovertemplate: 'Commit: %{text}<br>Date: %{x}<br>SLOC: %{y}<extra></extra>'
      };
      var trace2 = {
        x: dates,
        y: series2,
        mode: 'lines+markers',
        name: 'Non-testing Library Code (lib/)',
        text: commitHashes,
        hovertemplate: 'Commit: %{text}<br>Date: %{x}<br>SLOC: %{y}<extra></extra>'
      };
      var trace3 = {
        x: dates,
        y: series3,
        mode: 'lines+markers',
        name: 'Effective Compiled Code (out/)',
        text: commitHashes,
        hovertemplate: 'Commit: %{text}<br>Date: %{x}<br>SLOC: %{y}<extra></extra>'
      };

      var plotData = [trace1, trace2, trace3];
      var layout = {
        title: 'Git History SLOC Progression',
        xaxis: { title: 'Commit Date' },
        yaxis: { title: 'SLOC' },
        hovermode: 'closest'
      };

      Plotly.newPlot('chart', plotData, layout);
    </script>
  </body>
</html>
  `;
}

/**
 * Main execution function:
 *  - Retrieves commit history.
 *  - Processes each commit.
 *  - Generates the HTML visualization.
 */
async function main() {
  console.log("Gathering git commit history...");
  const commits = await getCommits();
  console.log(`Found ${commits.length} commits.`);

  const results = [];
  for (const commit of commits) {
    console.log(`Processing commit ${commit.hash}...`);
    const res = await processCommit(commit);
    if (res) {
      results.push(res);
    }
  }

  // Generate and write the HTML visualization.
  const htmlContent = generateHTML(results);
  const outputPath = path.join(process.cwd(), "git-sloc-history.html");
  fs.writeFileSync(outputPath, htmlContent);
  console.log(`Visualization generated: ${outputPath}`);
}

main().catch((err) => {
  console.error("Error:", err);
  process.exit(1);
});

File: dev_tools/git_history.rb
#!/usr/bin/env ruby
#
# A cross-platform Ruby re-implementation of the Node.js git-sloc-history script.
#
# It gathers commit history, archives and extracts each commit,
# runs "npx node-sloc" on various subdirectories,
# caches commit results (in "git_sloc_cache.json" next to this script),
# and generates an HTML visualization (with Plotly charts and a daily summary table).
#
# The daily table now shows “SLOC with tests” (all lib/ code) vs.
# “SLOC without tests” (lib/ with test files ignored).
#
require 'json'
require 'fileutils'
require 'tmpdir'
require 'open3'
require 'time'
require 'rubygems/package'

# Escape HTML special characters
def escape_html(str)
  str.to_s.gsub('&', '&amp;')
          .gsub('<', '&lt;')
          .gsub('>', '&gt;')
          .gsub('"', '&quot;')
          .gsub("'", '&#39;')
end

# Run a shell command in a given directory and return stdout.
def run_command(cmd, cwd = '.')
  stdout, stderr, status = Open3.capture3(cmd, chdir: cwd)
  raise "Command failed: #{cmd}\n#{stderr}" unless status.success?
  stdout
end

# Get commit history from git.
def get_commits
  cmd = 'git log --reverse "--pretty=format:%H^^^%ct^^^%s"'
  output = run_command(cmd)
  output.strip.split("\n").map do |line|
    parts = line.split('^^^')
    {
      'hash' => parts[0],
      'timestamp' => parts[1].to_i,
      'message' => parts[2..-1].join('^^^').strip
    }
  end
end

# Extract a tar archive using Ruby’s built-in tar reader.
def extract_tar(archive_path, destination)
  File.open(archive_path, "rb") do |file|
    Gem::Package::TarReader.new(file) do |tar|
      tar.each do |entry|
        dest = File.join(destination, entry.full_name)
        if entry.directory?
          FileUtils.mkdir_p(dest)
        else
          FileUtils.mkdir_p(File.dirname(dest))
          File.open(dest, "wb") { |f| f.write(entry.read) }
        end
      end
    end
  end
end

# Run node-sloc with the given arguments (relative to cwd) and parse its output.
def run_node_sloc(cwd, args_array)
  command = "npx node-sloc #{args_array.join(' ')}"
  output = run_command(command, cwd)
  result = { 'sloc' => 0, 'comments' => 0, 'blank' => 0, 'total' => 0 }
  result['sloc']     = output[/\|\s*SLOC\s*\|\s*(\d+)/, 1].to_i
  result['comments'] = output[/\|\s*Lines of comments\s*\|\s*(\d+)/, 1].to_i
  result['blank']    = output[/\|\s*Blank lines\s*\|\s*(\d+)/, 1].to_i
  result['total']    = output[/\|\s*Total LOC\s*\|\s*(\d+)/, 1].to_i
  result
end

# Process a single commit:
#   - Archive the commit into a tar file.
#   - Extract it to a temporary directory.
#   - Run three node-sloc commands on lib/ and out/ directories.
#   - Return a commit hash with added series stats.
def process_commit(commit)
  Dir.mktmpdir("git-sloc-") do |tmp_dir|
    archive_path = File.join(tmp_dir, "archive.tar")
    begin
      run_command("git archive #{commit['hash']} --format=tar -o \"#{archive_path}\"")
    rescue => e
      warn "Error archiving commit #{commit['hash']}: #{e}"
      return nil
    end

    begin
      extract_tar(archive_path, tmp_dir)
    rescue => e
      warn "Error extracting commit #{commit['hash']}: #{e}"
      return nil
    end

    # Run node-sloc on:
    #   series1: lib/ (all code, i.e. SLOC with tests)
    #   series2: lib/ ignoring test files (SLOC without tests)
    #   series3: out/ ignoring test files (compiled output)
    series1_stats = run_node_sloc('.', [File.join(tmp_dir, 'lib/')])
    series2_stats = run_node_sloc('.', [File.join(tmp_dir, 'lib/'), '--ignore-paths', '"**/*.test.ts"'])
    series3_stats = run_node_sloc('.', [File.join(tmp_dir, 'out/'), '--ignore-paths', '"**/*.test.js"'])

    commit.merge({
      'series1' => series1_stats,
      'series2' => series2_stats,
      'series3' => series3_stats
    })
  end
end

# Generate the HTML output (with Plotly charts and a daily summary table).
def generate_html(data)
  # Prepare arrays for charting.
  dates         = data.map { |d| Time.at(d['timestamp']).utc.iso8601.split('T')[0] }
  series1_sloc  = data.map { |d| d['series1']['sloc'] }
  series2_sloc  = data.map { |d| d['series2']['sloc'] }
  series3_sloc  = data.map { |d| d['series3']['sloc'] }
  messages      = data.map { |d| escape_html(d['message']) }
  indices       = (1..data.size).to_a

  # Build daily summaries for lib/:
  # "with tests" is series1.sloc and "without tests" is series2.sloc.
  daily_commits = {}
  data.each do |commit|
    day = Time.at(commit['timestamp']).utc.iso8601.split('T')[0]
    # Overwrite to keep the last commit of each day.
    daily_commits[day] = {
      'with_tests'    => commit['series1']['sloc'],
      'without_tests' => commit['series2']['sloc']
    }
  end
  sorted_days = daily_commits.keys.sort
  prev_with = 0
  prev_without = 0
  daily_rows = sorted_days.map do |day|
    with_tests    = daily_commits[day]['with_tests']
    without_tests = daily_commits[day]['without_tests']
    diff_with     = with_tests - prev_with
    diff_without  = without_tests - prev_without
    prev_with = with_tests
    prev_without = without_tests
    "<tr>
      <td>#{day}</td>
      <td>#{with_tests}</td>
      <td>#{diff_with}</td>
      <td>#{without_tests}</td>
      <td>#{diff_without}</td>
    </tr>"
  end.join("\n")

  <<~HTML
  <!DOCTYPE html>
  <html>
    <head>
      <meta charset="utf-8">
      <title>Git History SLOC Progression</title>
      <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
      <style>
        body { font-family: Arial, sans-serif; margin: 40px; }
        h1, h2 { text-align: center; }
        #chart-date, #chart-index {
          width: 100%;
          height: 600px;
          margin-bottom: 50px;
        }
        table {
          width: 100%;
          border-collapse: collapse;
          margin-top: 40px;
        }
        th, td {
          border: 1px solid #ccc;
          padding: 6px 10px;
          text-align: left;
        }
        th {
          background: #f9f9f9;
        }
      </style>
    </head>
    <body>
      <h1>Git History SLOC Progression</h1>

      <!-- Chart 1: X-axis = Dates -->
      <div id="chart-date"></div>

      <!-- Chart 2: X-axis = Commit # -->
      <div id="chart-index"></div>

      <!-- Daily Summaries for lib/:
           Showing end-of-day SLOC (with tests) vs.
           SLOC (without tests) plus the daily differences. -->
      <h2>Daily Summaries (lib/)</h2>
      <table>
        <thead>
          <tr>
            <th>Date</th>
            <th>End-of-Day SLOC (with tests)</th>
            <th>Daily Lines (with tests)</th>
            <th>End-of-Day SLOC (without tests)</th>
            <th>Daily Lines (without tests)</th>
          </tr>
        </thead>
        <tbody>
          #{daily_rows}
        </tbody>
      </table>

      <script>
        // Data for charts
        var dates = #{dates.to_json};
        var indices = #{indices.to_json};
        var series1 = #{series1_sloc.to_json};
        var series2 = #{series2_sloc.to_json};
        var series3 = #{series3_sloc.to_json};
        var messages = #{messages.to_json};

        // Chart 1: by commit date
        var trace1_date = {
          x: dates,
          y: series1,
          mode: 'lines+markers',
          name: 'All Code (lib/)',
          text: messages,
          hovertemplate: 'Message: %{text}<br>Date: %{x}<br>SLOC: %{y}<extra></extra>'
        };
        var trace2_date = {
          x: dates,
          y: series2,
          mode: 'lines+markers',
          name: 'Non-testing (lib/)',
          text: messages,
          hovertemplate: 'Message: %{text}<br>Date: %{x}<br>SLOC: %{y}<extra></extra>'
        };
        var trace3_date = {
          x: dates,
          y: series3,
          mode: 'lines+markers',
          name: 'Compiled (out/)',
          text: messages,
          hovertemplate: 'Message: %{text}<br>Date: %{x}<br>SLOC: %{y}<extra></extra>'
        };

        var layout_date = {
          title: 'SLOC vs Commit Date',
          xaxis: { title: 'Commit Date' },
          yaxis: { title: 'SLOC' },
          hovermode: 'closest'
        };
        Plotly.newPlot('chart-date', [trace1_date, trace2_date, trace3_date], layout_date);

        // Chart 2: by commit index
        var trace1_index = {
          x: indices,
          y: series1,
          mode: 'lines+markers',
          name: 'All Code (lib/)',
          text: messages,
          hovertemplate: 'Message: %{text}<br>Commit # %{x}<br>SLOC: %{y}<extra></extra>'
        };
        var trace2_index = {
          x: indices,
          y: series2,
          mode: 'lines+markers',
          name: 'Non-testing (lib/)',
          text: messages,
          hovertemplate: 'Message: %{text}<br>Commit # %{x}<br>SLOC: %{y}<extra></extra>'
        };
        var trace3_index = {
          x: indices,
          y: series3,
          mode: 'lines+markers',
          name: 'Compiled (out/)',
          text: messages,
          hovertemplate: 'Message: %{text}<br>Commit # %{x}<br>SLOC: %{y}<extra></extra>'
        };

        var layout_index = {
          title: 'SLOC vs Commit Number',
          xaxis: { title: 'Commit # (chronological)' },
          yaxis: { title: 'SLOC' },
          hovermode: 'closest'
        };
        Plotly.newPlot('chart-index', [trace1_index, trace2_index, trace3_index], layout_index);
      </script>
    </body>
  </html>
  HTML
end

# Main execution
def main
  puts "Loading cache..."
  script_dir = File.dirname(__FILE__)
  cache_file = File.join(script_dir, "~git_sloc_cache.json")
  cache = File.exist?(cache_file) ? JSON.parse(File.read(cache_file)) : {}

  puts "Gathering git commit history..."
  commits = get_commits
  puts "Found #{commits.size} commits."
  results = []
  commits.each_with_index do |commit, idx|
    if cache.key?(commit['hash'])
      puts "Using cached data for commit #{idx + 1}/#{commits.size}: #{commit['hash']}"
      results << cache[commit['hash']]
    else
      puts "Processing commit #{idx + 1}/#{commits.size}: #{commit['hash']}"
      res = process_commit(commit)
      if res
        results << res
        cache[commit['hash']] = res
      end
    end
  end

  # Write updated cache next to the script.
  File.write(cache_file, JSON.pretty_generate(cache))
  puts "Cache updated."

  html_content = generate_html(results)
  output_path = File.join(Dir.pwd, "~git-sloc-history.html")
  File.write(output_path, html_content)
  puts "Visualization generated: #{output_path}"
end

begin
  main
rescue => e
  warn "Error: #{e}"
  exit 1
end
